### PR TITLE
feat(repository): switched to using go-mmap for indexes

### PR DESCRIPTION
**Pull request by jkowalski**
Previous mmap implementation only allowed io.ReaderAt API
this one allows direct access to the underlying bytes which helps
remove a bunch of buffers, copying and allocation in index parsing.

### CODE REVIEW

File: repo/content/index/index.go [hunk 1]
```diff
@@ -26,33 +25,40 @@ type Index interface {
...
 		return nil, errors.Errorf("invalid header format: %v", h.version)
 	}
 }
 
-func readAtAll(ra io.ReaderAt, p []byte, offset int64) error {
-	n, err := ra.ReadAt(p, offset)
-	if n != len(p) {
-		return errors.Errorf("incomplete read at offset %v, got %v bytes, expected %v", offset, n, len(p))
-	}
+func safeSlice(data []byte, offset int64, length int) (v []byte, err error) {
```
Comment (julio-lopez) [inline repo/content/index/index.go:46]: what are the semantics of the `data` slice returned by the mmap library? Do either `cap(data)` or `len(data)` reflect the addressable region?

Are there errors other than out-of-bounds errors that need to be considered?
If that's the case, then the error used for recovery may be misleading.

File: repo/content/index/index.go [hunk 2]
```diff
@@ -26,33 +25,40 @@ type Index interface {
...
+	defer func() {
+		if recover() != nil {
+			err = errors.Errorf("invalid slice offset=%v, length=%v, data=%v bytes", offset, length, len(data))
```
Comment (julio-lopez) [inline repo/content/index/index.go:49]: Explicitly return a `nil []byte`  ?

```suggestion
			v = nil
			err = errors.Errorf("invalid slice offset=%v, length=%v, data=%v bytes", offset, length, len(data))
```

**Review by julio-lopez — APPROVED** [review]

File: repo/content/index/index.go (hunk 1 above)
Comment (jkowalski) [reply repo/content/index/index.go:46]: the slice is sized to the length of memory-mapped segment, so the bounds checker than runs when reslicing will do the right thing.

File: repo/content/committed_content_index_disk_cache.go [hunk 3]
```diff
@@ -68,10 +68,26 @@ func (c *diskCommittedContentIndexCache) mmapOpenWithRetry(path string) (*mmap.R
...
 		time.Sleep(nextDelay)
 		nextDelay *= 2
-		f, err = mmap.Open(path)
+
+		f, err = os.Open(path) // nolint:gosec
 	}
 
-	return f, errors.Wrap(err, "mmap() error")
+	mm, err := mmap.Map(f, mmap.RDONLY, 0)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (jkowalski) [reply repo/content/committed_content_index_disk_cache.go:76]: done.

File: repo/content/index/index.go (hunk 1 above)
Comment (julio-lopez) [reply repo/content/index/index.go:46]: In that case, would checking the bounds work instead of trying to `panic` and `recover` ?

**Pull request merged**